File: 백준/Gold/16472. 고냥이/고냥이.cc
#include <iostream>
using namespace std;

int n;
string s;

int main()
{
	cin >> n >> s;
	int left = 0, right = 0;
	int alpha[26] = {
		0,
	};
	int ans = 1, cnt = 1;
	alpha[s[0] - 'a'] = 1;
	while (left <= right && right < s.length())
	{
		while (1)
		{
			right++;
			if (!alpha[s[right] - 'a'])
			{
				cnt++;
			}
			alpha[s[right] - 'a']++;

			if (cnt > n || right >= s.length())
			{
				break;
			}

			ans = max(ans, right - left + 1);
		}

		while (cnt > n)
		{
			alpha[s[left] - 'a']--;
			left++;
			if (alpha[s[left - 1] - 'a'] == 0)
			{
				cnt--;
			}

			if (cnt <= n)
			{
				ans = max(ans, right - left + 1);
				break;
			}
		}
	}

	cout << ans;
}

File: 백준/Gold/16472. 고냥이/고냥이.java
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Set;
import java.util.StringTokenizer;

public class Main {

	public static void main(String[] args) throws Exception {
		var br = new BufferedReader(new InputStreamReader(System.in));

		int n = Integer.parseInt(br.readLine());
		String s = br.readLine();

		Map<Character, Integer> map = new HashMap<>();
		int left = 0, right = -1, ans = 0;
		while (right < s.length() - 1) {
			right++;
			char c = s.charAt(right);
			map.put(c, map.getOrDefault(c, 0) + 1);
			while (map.size() > n) {
				char l = s.charAt(left);
				map.put(l, map.get(l) - 1);
				if (map.get(l) == 0) {
					map.remove(l);
				}
				left++;
			}
			ans = Math.max(ans, right - left + 1);
		}

		System.out.println(ans);
	}
}